### ISSUE TITLE
useVModel(s) should optionally be backwards compatible.

### ISSUE THREAD

**Issue by JessicaSachs** [vueuse/vueuse#2317]
### Clear and concise description of the problem

I would like to create a runtime bridge for v-model to work with Vue 2.X and Vue 3 components. [Vue Bridge wants to solve this problem](https://vue-bridge-docs.netlify.app/reference/runtime.html), but is not currently publicized heavily as it is under construction. 

I believe VueUse would have better adoption + usage due to its popularity and it already depends on vue-demi.

Is this out of scope?

### Suggested solution

In VueUse core, we could extend useVModel to take in an option for backwards compatibility.

**As an aside:**
I would like Volar + vue-tsc to help create a version of `defineProps` and `defineEmits` that would support this macro. Is an experimental feature possible @johnsoncodehk? I think I will create an issue in either project (vue-tsc || Volar) regardless of if we end up adding this functionality to VueUse.

### Alternative

- We could create a separate function so that useVModel is not weighed down by a heavier API surface area.
- We could decide this is out of scope for VueUse

### Validations

- [X] Follow our [Code of Conduct](https://github.com/vueuse/vueuse/blob/main/CODE_OF_CONDUCT.md)
- [X] Read the [Contributing Guidelines](https://github.com/vueuse/vueuse/blob/main/CONTRIBUTING.md).
- [X] Read the [docs](https://vueuse.org/guide).
- [X] Check that there isn't already an issue that request the same feature to avoid creating a duplicate.

**Comment by sxzz** [comment]
I think `defineModel` of `unplugin-vue-macros` is what exactly you described.

https://github.com/sxzz/unplugin-vue-macros#definemodel

**Comment by JessicaSachs** [comment]
Yes! **It's _almost there_, but the open issues block me from using it.** Can you double check my usage?

**Pros:**
1. Not having to do a build-time transform or explicit runtime logic would be nice!!!
2. Unplugin ecosystem makes this accessible for lots of users :+1:

**Blockers:**
1. Does not seem to work with Vue 2.7. Can you check if I am doing it right?
    1. Here is a video and <a href="https://github.com/JessicaSachs/petite/compare/develop...v-model-support">reproduction</a> and a <a href="https://user-images.githubusercontent.com/2801156/195407170-c22ec7d5-967d-45a8-960d-21fb13904f42.mov">High-resolution video download</a><br/><details><summary>Low-res gif</summary><img src="https://user-images.githubusercontent.com/2801156/195408674-53abb1f8-5500-4150-8fca-313cd5c402d4.gif" /></details>
2. TypeScript / Volar is not supported (written in the known issues, but this breaks all prop-type checking).
4. `withDefaults` not supported
5. Not discoverable(yet!)

**Comment by JessicaSachs** [comment]
Just a thought: instead of solving the build-time and blocking progress on typescript tooling issues, we could decide to put this in the VueUse runtime as it would still be accessible to users using VueUse without a bundler.

Once we fix the blockers, maybe we can use the VueUse docs to link out to `unplugin-vue-macros`?

**Comment by sxzz** [comment]
FYI: https://discord.com/channels/793943652350427136/1029817571916128326

**Comment by JessicaSachs** [comment]
TL;DR: @sxzz will implement support for backwards-compatible emits + modelValue alongside the necessary Volar + vue-tsc changes. With these things together, we will have support for backwards-compatible `<script setup lang="ts">`

![image](https://user-images.githubusercontent.com/2801156/195428600-885afb00-54c4-46be-a879-01e126ff80e8.png)

And under the hood, this issue will emit to both methods. 

```ts
// They're refs.
const { title, modelValue, value } = defineModel<{
  title: string
  value: string
  modelValue: string
}>()

const update = () => {
  title.value += ' Updated!'
  value.value += ' Updated!'
  modelValue.value += ' Updated!'
  debugger
}
```

**Comment by sxzz** [comment]
In summary. I'll work on `defineModel` version that doesn't need Reactivity Transform, plus Volar plugin (it's done for development, but need to change a little for supported no Reactivity Transform).

This version will depend on @vueuse/core, and be implemented by useVModel, so it will also support all options of it (UseVModelOptions). For example `passive`, a pretty useful option. You can refer to VueUse docs for more detailed descriptions. https://vueuse.org/core/usevmodel/#usevmodel . At the first, I didn't make this version is that I don't want to depend on VueUse. XD

As for Vue 3, it works now. If you have any issues with Vue 3, feel free to tell me.

**Comment by sxzz** [comment]
@JessicaSachs Hi, I already added the support of Vue 2 and the unified prop name. It will transform Vue 3 code to Vue 2 seamless. I also already refactored the code of your project. So just go to check code~
https://github.com/JessicaSachs/petite/compare/v-model-support...sxzz:petite:v-model-support

**Comment by JessicaSachs** [comment]
I'll reopen this if it's not what I need.
